app/src/main/java/com/example/taskyy/data/repositories/AuthRepositoryImpl.kt

```diff
@@ -59,7 +65,26 @@ class AuthRepositoryImpl @Inject constructor(
         return hasLowerCase && hasUpperCase && hasDigit && hasValidLength
     }
 
-    override fun login(login: Login): Call<LoginUserResponse> {
-        return retrofit.loginUser(login)
+    override fun login(login: Login): Result<LoginUserResponse> {
+        val loginUser = retrofit.loginUser(login)
+        var returnStatement: Result<LoginUserResponse>
+        loginUser.enqueue(object : Callback<LoginUserResponse> {
+            override fun onResponse(
+                call: Call<LoginUserResponse>,
+                response: Response<LoginUserResponse>
+            ) {
+                val responseInfo: LoginUserResponse = response.body() as LoginUserResponse
+                addTokenAndIdToDatabase(responseInfo, login.email)
+                if(response.isSuccessful){
+                    returnStatement = Result.success()
+                } else {
+                    returnStatement = Result.failure(exception = Throwable())
+                }
+            }
+
+            override fun onFailure(call: Call<LoginUserResponse>, t: Throwable) {
+            }
+        })
```

Comment:
You're using Retrofit's callback mechanism here, which is all fine but doesn't work well with coroutines. Instead, the easiest is to make all your Retrofit functions `suspend fun`:
```
@POST("/login)
suspend fun login(@Body body: LoginRequest): UserDto
```
Then you can call them in your repo just like this:
```
override suspend fun login(login: Login): Result<LoginUserResponse> {
    return try {
        val user = retrofit.login(...)
        Result.success(user.toUser())
    } catch(e: HttpException) {
        Result.failure(e)
    } catch(e: IOException) {
        Result.failure(e)
    }
```
I personally use a custom `Result` wrapper class which is a bit more flexible and optimized than the Kotlin one here, but at this point we should focus on you understanding the core mechanism behind this. Then we can later get into the little optimizations :)

Reply:
This is confusing for me. Why would I return a UserDto (which currently doesn't exist) from the login request? Also, in the repo, why am I attempting to pass user.toUser() into the Result.success()? I don't understand what I'm doing here. I created the LoginResponse class which I assumed would be used as the return type from the login call

Reply:
I think I commented on that earlier, but in the type of architecture you chose (presentation/domain/data layers), communication always points inwards at domain. That means both presentation and data layers are allowed to access classes from domain, but domain is not allowed to access any other layer. It also shouldn't know where data comes from because it would then indirectly know something about the data layer. 

That's why we have both data models (Such as `UserDto` from the API or `UserEntity` from the DB) which we map to a domain model in the repository. That way, domain only knows its own model and not any others from the data layer (such as your `LoginUserResponse` class).

What I called `UserDto` is the same as your `LoginUserResponse` here, we just used different naming. Either way, to adhere to the architectural guidelines, you would need to convert the `LoginUserResponse` to a model from domain.